src/assets/TokenListController.ts

```diff
@@ -0,0 +1,270 @@
+   */
+  stop() {
+    this.stopPolling();
+  }
+
+  /**
+   * Prepare to discard this controller.
+   *
+   * This stops any active polling.
+   */
+  destroy() {
+    super.destroy();
+    this.stopPolling();
+  }
+
+  private stopPolling() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  /**
+   * Starts a new polling interval
+   */
+  private async startPolling(): Promise<void> {
+    await safelyExecute(() => this.fetchTokenList());
+    this.intervalId = setInterval(async () => {
+      await safelyExecute(() => this.fetchTokenList());
+    }, this.intervalDelay);
+  }
+
+  /**
+   * Fetching token list from the Token Service API
+   */
+  async fetchTokenList(): Promise<void> {
+    const releaseLock = await this.mutex.acquire();
+    try {
+      const tokensFromAPI: Token[] = await safelyExecute(() =>
+        this.fetchFromCache(),
+      );
+      const { tokensChainsCache } = this.state;
+      const tokenList: TokenMap = {};
+
+      // filtering out tokens with less than 2 occurences
+      const filteredTokenList = tokensFromAPI.filter(
+        (token) => token.occurrences >= 2,
+      );
+      // removing the tokens with symbol conflicts
+      const symbolsList = filteredTokenList.map((token) => token.symbol);
+      const duplicateSymbols = [
+        ...new Set(
+          symbolsList.filter(
+            (symbol, index) => symbolsList.indexOf(symbol) !== index,
+          ),
+        ),
+      ];
+      const uniqueTokenList = filteredTokenList.filter(
+        (token) => !duplicateSymbols.includes(token.name),
+      );
+      for (const token of uniqueTokenList) {
+        tokenList[token.address] = token;
+      }
+      this.update(() => {
+        return {
+          tokenList,
+          tokensChainsCache,
+        };
+      });
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Checks if the Cache timestamp is valid,
+   *  if yes data in cache will be returned
+   *  otherwise a call to the API service will be made.
+   * @returns Promise that resolves into a TokenList
+   */
+  async fetchFromCache(): Promise<Token[]> {
+    const { tokensChainsCache, ...tokensData }: TokenListState = this.state;
+    const dataCache = tokensChainsCache[this.chainId];
+    const now = Date.now();
+    if (dataCache && now - dataCache.timeStamp < this.cacheRefreshThreshold) {
+      return dataCache.data;
+    }
+    const tokenList: Token[] = await safelyExecute(() =>
+      fetchTokenList(this.chainId),
+    );
+    const updatedTokensChainsCache = {
+      ...tokensChainsCache,
+      [this.chainId]: {
+        timeStamp: Date.now(),
+        data: tokenList,
+      },
+    };
+    this.update(() => {
+      return {
+        ...tokensData,
+        tokensChainsCache: updatedTokensChainsCache,
+      };
+    });
+    return tokenList;
+  }
+
+  /**
+   * Calls the API to sync the tokens in the token service
+   */
+  async syncTokens(): Promise<void> {
```

Comment:
Correct me if I'm wrong: If we left the default interval and threshold values alone, every time we poll, we should be hitting the API and not cache. I can imagine a case where if we lowered the default interval + called syncTokens, we may still receive cached data since the default threshold may not have expired yet. Btw, this shouldn't be a problem since at some point, that threshold will expire and polling will eventually hit the API. However, would it make sense to invalidate the cache every time syncTokens gets called so that the next poll can always guarantee fresh data ASAP?

Reply:
I have updated the code based on the suggestion.